File: src/integrations/agentic-data-kernel.ts
import { WorldCutIntegrationError } from "../errors.js";
import { snapshotJsonData } from "../canonical.js";
import type {
  CoherenceContract,
  DependencyWitness,
  JsonValue,
  Observation,
  ResourceIdentity,
  WitnessProvenance,
} from "../types.js";
import { verifyDecisionContract } from "../verifier.js";

export interface AgenticDataAssertionLike {
  tenantId: string;
  assertionId: string;
  object: JsonValue;
  validFrom: string;
  validTo: string | null;
  systemFrom: string;
  systemTo: string | null;
  status:
    | "active"
    | "disputed"
    | "superseded"
    | "expired"
    | "quarantined"
    | "deleted";
  basis: JsonValue | null;
}

export interface AgenticDataResolutionLike {
  status: "known" | "unknown" | "conflicted" | "resolved_with_conflict";
  selected: AgenticDataAssertionLike | null;
  validAt: string;
  systemAt: string;
}

export interface AgenticDataAdapterOptions {
  allowResolvedWithConflict?: boolean;
}

interface WorldCutBasis {
  protocolVersion: "0.1";
  role: string;
  resource: ResourceIdentity;
  provenance: WitnessProvenance;
  version?: string;
  dependencies?: DependencyWitness[];
  acquisitionCost?: number;
}

function invalid(message: string): never {
  throw new WorldCutIntegrationError(
    "WORLDCUT_ADK_RESOLUTION_INVALID",
    message,
  );
}

function record(value: unknown, field: string): Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    return invalid(`${field} must be an object`);
  }
  return value as Record<string, unknown>;
}

function timestamp(value: unknown, field: string): number {
  if (typeof value !== "string") {
    return invalid(`${field} must be a normalized timestamp`);
  }
  const parsed = Date.parse(value);
  if (!Number.isFinite(parsed) || new Date(parsed).toISOString() !== value) {
    return invalid(`${field} must be normalized ISO-8601 UTC`);
  }
  return parsed;
}

function activeAt(
  start: string,
  end: string | null,
  at: string,
  field: string,
): boolean {
  const startTime = timestamp(start, `${field}.from`);
  const atTime = timestamp(at, `${field}.at`);
  const endTime = end === null ? Number.POSITIVE_INFINITY : timestamp(end, `${field}.to`);
  return startTime <= atTime && atTime < endTime;
}

function worldCutBasis(assertion: AgenticDataAssertionLike): WorldCutBasis {
  const basis = record(assertion.basis, "assertion.basis");
  const candidate = record(basis.worldcut, "assertion.basis.worldcut");
  const allowedFields = new Set([
    "protocolVersion",
    "role",
    "resource",
    "provenance",
    "version",
    "dependencies",
    "acquisitionCost",
  ]);
  const unknownFields = Object.keys(candidate).filter(
    (field) => !allowedFields.has(field),
  );
  if (unknownFields.length > 0) {
    return invalid(
      `assertion.basis.worldcut contains unsupported field(s): ${unknownFields.join(", ")}`,
    );
  }
  if (candidate.protocolVersion !== "0.1") {
    return invalid("assertion.basis.worldcut.protocolVersion must equal 0.1");
  }
  if (typeof candidate.role !== "string" || candidate.role.length === 0) {
    return invalid("assertion.basis.worldcut.role must be a non-empty string");
  }
  const resource = record(
    candidate.resource,
    "assertion.basis.worldcut.resource",
  );
  for (const field of ["provider", "account", "kind", "key"]) {
    if (typeof resource[field] !== "string" || resource[field].length === 0) {
      return invalid(
        `assertion.basis.worldcut.resource.${field} must be a non-empty string`,
      );
    }
  }
  const provenanceValues = new Set([
    "provider_asserted",
    "client_observed",
    "derived",
    "operator_supplied",
  ]);
  if (
    typeof candidate.provenance !== "string" ||
    !provenanceValues.has(candidate.provenance)
  ) {
    return invalid("assertion.basis.worldcut.provenance is unsupported");
  }
  if (
    candidate.acquisitionCost !== undefined &&
    (typeof candidate.acquisitionCost !== "number" ||
      !Number.isFinite(candidate.acquisitionCost) ||
      candidate.acquisitionCost < 0)
  ) {
    return invalid(
      "assertion.basis.worldcut.acquisitionCost must be non-negative",
    );
  }
  const snapshot = snapshotJsonData(candidate) as WorldCutBasis;
  if (
    snapshot.dependencies !== undefined &&
    !Array.isArray(snapshot.dependencies)
  ) {
    return invalid(
      "assertion.basis.worldcut.dependencies must be an array",
    );
  }
  for (const dependency of snapshot.dependencies ?? []) {
    const dependencyRecord = record(
      dependency,
      "assertion.basis.worldcut.dependencies[]",
    );
    const dependencyResource = record(
      dependencyRecord.resource,
      "assertion.basis.worldcut.dependencies[].resource",
    );
    if (dependencyResource.account !== assertion.tenantId) {
      return invalid(
        "Every WorldCut dependency resource account must equal the Agentic Data Kernel tenantId",
      );
    }
  }
  return snapshot;
}

function adaptAgenticDataResolution(
  resolutionInput: AgenticDataResolutionLike,
  options: AgenticDataAdapterOptions = {},
): Observation {
  const resolution = snapshotJsonData(
    resolutionInput,
    "resolution",
  ) as AgenticDataResolutionLike;
  if (
    ![
      "known",
      "unknown",
      "conflicted",
      "resolved_with_conflict",
    ].includes(resolution.status)
  ) {
    return invalid(
      `Agentic Data Kernel resolution status ${String(resolution.status)} is unsupported`,
    );
  }
  if (resolution.status === "unknown" || resolution.status === "conflicted") {
    return invalid(
      `Agentic Data Kernel resolution status ${resolution.status} cannot authorize an observation`,
    );
  }
  if (
    resolution.status === "resolved_with_conflict" &&
    !options.allowResolvedWithConflict
  ) {
    return invalid(
      "resolved_with_conflict requires allowResolvedWithConflict: true",
    );
  }
  const assertion = resolution.selected;
  if (!assertion) {
    return invalid("Agentic Data Kernel resolution has no selected assertion");
  }
  if (assertion.status !== "active") {
    return invalid(
      `Agentic Data Kernel assertion status ${assertion.status} is not eligible`,
    );
  }
  if (
    !activeAt(
      assertion.systemFrom,
      assertion.systemTo,
      resolution.systemAt,
      "assertion.systemTime",
    )
  ) {
    return invalid(
      "Agentic Data Kernel assertion is not system-valid at resolution.systemAt",
    );
  }
  if (
    !activeAt(
      assertion.validFrom,
      assertion.validTo,
      resolution.validAt,
      "assertion.validTime",
    )
  ) {
    return invalid(
      "Agentic Data Kernel assertion is not business-valid at resolution.validAt",
    );
  }

  const basis = worldCutBasis(assertion);
  if (basis.resource.account !== assertion.tenantId) {
    return invalid(
      "WorldCut resource account must equal the Agentic Data Kernel tenantId",
    );
  }

  const observation: Observation = {
    id: `adk-${assertion.assertionId}`,
    role: basis.role,
    resource: basis.resource,
    value: assertion.object,
    observedAt: assertion.systemFrom,
    acquisitionCost: basis.acquisitionCost ?? 1,
    witness: {
      provenance: basis.provenance,
      ...(basis.version ? { version: basis.version } : {}),
      validity: {
        from: assertion.validFrom,
        until: assertion.validTo,
      },
      ...(basis.dependencies
        ? { dependencies: basis.dependencies }
        : {}),
    },
  };
  const validationContract: CoherenceContract = {
    id: "agentic-data-kernel-observation-validation",
    version: "1",
    decisionTime: resolution.systemAt,
    assumptions: {
      clockModel: "trusted_normalized",
      intervalModel: "half_open",
      metadataModel: "honest_but_possibly_incomplete",
    },
    requirements: [
      {
        id: "selected-value-is-preserved",
        type: "value_equals",
        description: "The selected kernel value is preserved",
        role: observation.role,
        path: ["selected"],
        expected: observation.value,
      },
    ],
  };
  try {
    verifyDecisionContract({
      protocolVersion: "0.1",
      contract: validationContract,
      observations: [
        {
          ...observation,
          value: {
            selected: observation.value,
          },
        },
      ],
    });
  } catch (error) {
    throw new WorldCutIntegrationError(
      "WORLDCUT_ADK_RESOLUTION_INVALID",
      "Agentic Data Kernel WorldCut metadata is invalid",
      { cause: error },
    );
  }
  return observation;
}

export function observationFromAgenticDataResolution(
  resolutionInput: AgenticDataResolutionLike,
  options: AgenticDataAdapterOptions = {},
): Observation {
  try {
    return adaptAgenticDataResolution(resolutionInput, options);
  } catch (error) {
    if (error instanceof WorldCutIntegrationError) {
      throw error;
    }
    throw new WorldCutIntegrationError(
      "WORLDCUT_ADK_RESOLUTION_INVALID",
      "Agentic Data Kernel resolution metadata is invalid",
      { cause: error },
    );
  }
}

File: ports/go/integrations/agenticdatakernel/agenticdatakernel.go
// Package agenticdatakernel adapts resolved Agentic Data Kernel assertions
// into WorldCut observations.
//
// The adapter is structural: it declares only the fields WorldCut needs and
// takes no runtime dependency on the kernel. Every rejection uses the stable
// WORLDCUT_ADK_RESOLUTION_INVALID code and fails closed.
package agenticdatakernel

import (
	"fmt"
	"sort"
	"time"

	worldcut "github.com/Jason-Doyle/WorldCut/ports/go"
)

// Assertion is the structural subset of a kernel assertion WorldCut consumes.
type Assertion struct {
	// TenantID owns the assertion. Every WorldCut resource account and
	// dependency resource account must equal it.
	TenantID string
	// AssertionID identifies the assertion and the derived observation.
	AssertionID string
	// Object is the asserted JSON value.
	Object any
	// ValidFrom and ValidTo bound business validity. A nil ValidTo is open.
	ValidFrom string
	ValidTo   *string
	// SystemFrom and SystemTo bound system validity. A nil SystemTo is open.
	SystemFrom string
	SystemTo   *string
	// Status is the assertion lifecycle state. Only "active" is eligible.
	Status string
	// Basis carries the namespaced basis.worldcut metadata.
	Basis any
}

// Resolution is the structural subset of a kernel resolution result.
type Resolution struct {
	// Status is one of known, unknown, conflicted, or resolved_with_conflict.
	Status string
	// Selected is the assertion the kernel selected, when one exists.
	Selected *Assertion
	// ValidAt and SystemAt are the bitemporal coordinates of the resolution.
	ValidAt  string
	SystemAt string
}

// Options carries application policy for adapting a resolution.
type Options struct {
	// AllowResolvedWithConflict permits a resolution that preserved an
	// unresolved conflict. This is an explicit application policy decision.
	AllowResolvedWithConflict bool
}

var (
	resolutionStatuses = map[string]bool{
		"known":                  true,
		"unknown":                true,
		"conflicted":             true,
		"resolved_with_conflict": true,
	}
	provenanceValues = map[string]bool{
		"provider_asserted": true,
		"client_observed":   true,
		"derived":           true,
		"operator_supplied": true,
	}
	basisFields = map[string]bool{
		"protocolVersion": true,
		"role":            true,
		"resource":        true,
		"provenance":      true,
		"version":         true,
		"dependencies":    true,
		"acquisitionCost": true,
	}
	resourceFields = []string{"provider", "account", "kind", "key"}
)

func invalid(format string, arguments ...any) error {
	return worldcut.NewError(
		worldcut.ADKResolutionInvalidCode,
		fmt.Sprintf(format, arguments...),
	)
}

// copyText detaches an optional caller-owned string pointer so later
// assignment through that pointer cannot change a returned observation.
func copyText(value *string) *string {
	if value == nil {
		return nil
	}
	copied := *value
	return &copied
}

func wrap(message string, cause error) error {
	return worldcut.WrapError(worldcut.ADKResolutionInvalidCode, message, cause)
}

type worldCutBasis struct {
	role            string
	resource        worldcut.ResourceIdentity
	provenance      string
	version         *string
	dependencies    []worldcut.DependencyWitness
	acquisitionCost int64
}

// ObservationFromResolution converts an eligible kernel resolution into a
// validated WorldCut observation.
//
// The returned observation shares no memory with the caller's assertion
// object or basis metadata, and it has already passed the same strict
// protocol validation as a transported verification input.
func ObservationFromResolution(resolution Resolution, options Options) (worldcut.Observation, error) {
	observation, err := adapt(resolution, options)
	if err != nil {
		if worldcut.ErrorCode(err) == worldcut.ADKResolutionInvalidCode {
			return worldcut.Observation{}, err
		}
		return worldcut.Observation{}, wrap(
			"Agentic Data Kernel resolution metadata is invalid",
			err,
		)
	}
	return observation, nil
}

func adapt(resolution Resolution, options Options) (worldcut.Observation, error) {
	if !resolutionStatuses[resolution.Status] {
		return worldcut.Observation{}, invalid(
			"Agentic Data Kernel resolution status %s is unsupported",
			resolution.Status,
		)
	}
	if resolution.Status == "unknown" || resolution.Status == "conflicted" {
		return worldcut.Observation{}, invalid(
			"Agentic Data Kernel resolution status %s cannot authorize an observation",
			resolution.Status,
		)
	}
	if resolution.Status == "resolved_with_conflict" && !options.AllowResolvedWithConflict {
		return worldcut.Observation{}, invalid(
			"resolved_with_conflict requires AllowResolvedWithConflict: true",
		)
	}
	if resolution.Selected == nil {
		return worldcut.Observation{}, invalid(
			"Agentic Data Kernel resolution has no selected assertion",
		)
	}
	assertion := *resolution.Selected
	if assertion.Status != "active" {
		return worldcut.Observation{}, invalid(
			"Agentic Data Kernel assertion status %s is not eligible",
			assertion.Status,
		)
	}

	systemValid, err := activeAt(
		assertion.SystemFrom,
		assertion.SystemTo,
		resolution.SystemAt,
		"assertion.systemTime",
	)
	if err != nil {
		return worldcut.Observation{}, err
	}
	if !systemValid {
		return worldcut.Observation{}, invalid(
			"Agentic Data Kernel assertion is not system-valid at resolution.systemAt",
		)
	}
	businessValid, err := activeAt(
		assertion.ValidFrom,
		assertion.ValidTo,
		resolution.ValidAt,
		"assertion.validTime",
	)
	if err != nil {
		return worldcut.Observation{}, err
	}
	if !businessValid {
		return worldcut.Observation{}, invalid(
			"Agentic Data Kernel assertion is not business-valid at resolution.validAt",
		)
	}

	object, err := worldcut.SnapshotJSONValue(assertion.Object)
	if err != nil {
		return worldcut.Observation{}, wrap("assertion.object is not JSON data", err)
	}
	basis, err := worldCutBasisFrom(assertion)
	if err != nil {
		return worldcut.Observation{}, err
	}
	if basis.resource.Account != assertion.TenantID {
		return worldcut.Observation{}, invalid(
			"WorldCut resource account must equal the Agentic Data Kernel tenantId",
		)
	}

	observation := worldcut.Observation{
		ID:              "adk-" + assertion.AssertionID,
		Role:            basis.role,
		Resource:        basis.resource,
		Value:           object,
		ObservedAt:      assertion.SystemFrom,
		AcquisitionCost: basis.acquisitionCost,
		Witness: worldcut.ObservationWitness{
			Provenance: basis.provenance,
			Version:    basis.version,
			Validity: &worldcut.ValidityInterval{
				From:  assertion.ValidFrom,
				Until: copyText(assertion.ValidTo),
			},
			Dependencies: basis.dependencies,
		},
	}
	if err := validateObservation(observation, resolution.SystemAt); err != nil {
		return worldcut.Observation{}, wrap(
			"Agentic Data Kernel WorldCut metadata is invalid",
			err,
		)
	}
	return observation, nil
}

// validateObservation submits the adapted observation to the WorldCut engine
// so that construction cannot bypass protocol validation.
func validateObservation(observation worldcut.Observation, systemAt string) error {
	probe := observation
	probe.Value = map[string]any{"selected": observation.Value}
	_, err := worldcut.VerifyDecisionContract(worldcut.VerificationInput{
		ProtocolVersion: worldcut.ProtocolVersion,
		Contract: worldcut.Contract{
			ID:           "agentic-data-kernel-observation-validation",
			Version:      "1",
			DecisionTime: systemAt,
			Assumptions:  worldcut.SupportedAssumptions(),
			Requirements: []worldcut.Requirement{
				worldcut.NewValueEqualsRequirement(
					"selected-value-is-preserved",
					"The selected kernel value is preserved",
					observation.Role,
					[]string{"selected"},
					observation.Value,
				),
			},
		},
		Observations: []worldcut.Observation{probe},
	})
	return err
}

func timestamp(value, field string) (time.Time, error) {
	parsed, err := worldcut.ParseTimestamp(value)
	if err != nil {
		return time.Time{}, invalid("%s must be normalized ISO-8601 UTC", field)
	}
	return parsed, nil
}

func activeAt(start string, end *string, at, field string) (bool, error) {
	startTime, err := timestamp(start, field+".from")
	if err != nil {
		return false, err
	}
	atTime, err := timestamp(at, field+".at")
	if err != nil {
		return false, err
	}
	if end == nil {
		return !atTime.Before(startTime), nil
	}
	endTime, err := timestamp(*end, field+".to")
	if err != nil {
		return false, err
	}
	return !atTime.Before(startTime) && atTime.Before(endTime), nil
}

func worldCutBasisFrom(assertion Assertion) (worldCutBasis, error) {
	snapshot, err := worldcut.SnapshotJSONValue(assertion.Basis)
	if err != nil {
		return worldCutBasis{}, wrap("assertion.basis is not JSON data", err)
	}
	basisRecord, ok := snapshot.(map[string]any)
	if !ok {
		return worldCutBasis{}, invalid("assertion.basis must be an object")
	}
	candidate, ok := basisRecord["worldcut"].(map[string]any)
	if !ok {
		return worldCutBasis{}, invalid("assertion.basis.worldcut must be an object")
	}
	unknown := unsupportedFields(candidate, basisFields)
	if len(unknown) != 0 {
		return worldCutBasis{}, invalid(
			"assertion.basis.worldcut contains unsupported field(s): %v",
			unknown,
		)
	}
	if candidate["protocolVersion"] != worldcut.ProtocolVersion {
		return worldCutBasis{}, invalid(
			"assertion.basis.worldcut.protocolVersion must equal %s",
			worldcut.ProtocolVersion,
		)
	}
	role, ok := candidate["role"].(string)
	if !ok || role == "" {
		return worldCutBasis{}, invalid("assertion.basis.worldcut.role must be a non-empty string")
	}
	resource, err := resourceFrom(candidate["resource"], "assertion.basis.worldcut.resource")
	if err != nil {
		return worldCutBasis{}, err
	}
	provenance, ok := candidate["provenance"].(string)
	if !ok || !provenanceValues[provenance] {
		return worldCutBasis{}, invalid("assertion.basis.worldcut.provenance is unsupported")
	}
	var version *string
	if rawVersion, exists := candidate["version"]; exists {
		text, ok := rawVersion.(string)
		if !ok || text == "" {
			return worldCutBasis{}, invalid(
				"assertion.basis.worldcut.version must be a non-empty string",
			)
		}
		version = &text
	}
	acquisitionCost := int64(1)
	if rawCost, exists := candidate["acquisitionCost"]; exists {
		number, ok := rawCost.(float64)
		if !ok || number != float64(int64(number)) || number < 0 ||
			number > float64(worldcut.MaxAcquisitionCost) {
			return worldCutBasis{}, invalid(
				"assertion.basis.worldcut.acquisitionCost must be an integer between 0 and %d",
				worldcut.MaxAcquisitionCost,
			)
		}
		acquisitionCost = int64(number)
	}
	dependencies, err := dependenciesFrom(candidate, assertion.TenantID)
	if err != nil {
		return worldCutBasis{}, err
	}
	return worldCutBasis{
		role:            role,
		resource:        resource,
		provenance:      provenance,
		version:         version,
		dependencies:    dependencies,
		acquisitionCost: acquisitionCost,
	}, nil
}

func dependenciesFrom(candidate map[string]any, tenantID string) ([]worldcut.DependencyWitness, error) {
	rawDependencies, exists := candidate["dependencies"]
	if !exists {
		return nil, nil
	}
	values, ok := rawDependencies.([]any)
	if !ok {
		return nil, invalid("assertion.basis.worldcut.dependencies must be an array")
	}
	dependencies := make([]worldcut.DependencyWitness, 0, len(values))
	for _, value := range values {
		record, ok := value.(map[string]any)
		if !ok {
			return nil, invalid("assertion.basis.worldcut.dependencies[] must be an object")
		}
		resourceRecord, ok := record["resource"].(map[string]any)
		if !ok {
			return nil, invalid(
				"assertion.basis.worldcut.dependencies[].resource must be an object",
			)
		}
		if resourceRecord["account"] != tenantID {
			return nil, invalid(
				"Every WorldCut dependency resource account must equal the Agentic Data Kernel tenantId",
			)
		}
		unknown := unsupportedFields(record, map[string]bool{
			"name":       true,
			"resource":   true,
			"relation":   true,
			"version":    true,
			"provenance": true,
		})
		if len(unknown) != 0 {
			return nil, invalid(
				"assertion.basis.worldcut.dependencies[] contains unsupported field(s): %v",
				unknown,
			)
		}
		name, ok := record["name"].(string)
		if !ok || name == "" {
			return nil, invalid(
				"assertion.basis.worldcut.dependencies[].name must be a non-empty string",
			)
		}
		resource, err := resourceFrom(
			resourceRecord,
			"assertion.basis.worldcut.dependencies["+name+"].resource",
		)
		if err != nil {
			return nil, err
		}
		relation, ok := record["relation"].(string)
		if !ok || relation != "exact" {
			return nil, invalid(
				"assertion.basis.worldcut.dependencies[%s].relation is unsupported",
				name,
			)
		}
		provenance, ok := record["provenance"].(string)
		if !ok || !provenanceValues[provenance] {
			return nil, invalid(
				"assertion.basis.worldcut.dependencies[%s].provenance is unsupported",
				name,
			)
		}
		var version *string
		if rawVersion, exists := record["version"]; exists {
			text, ok := rawVersion.(string)
			if !ok || text == "" {
				return nil, invalid(
					"assertion.basis.worldcut.dependencies[%s].version must be a non-empty string",
					name,
				)
			}
			version = &text
		}
		dependencies = append(dependencies, worldcut.DependencyWitness{
			Name:       name,
			Resource:   resource,
			Relation:   relation,
			Version:    version,
			Provenance: provenance,
		})
	}
	return dependencies, nil
}

func resourceFrom(value any, field string) (worldcut.ResourceIdentity, error) {
	record, ok := value.(map[string]any)
	if !ok {
		return worldcut.ResourceIdentity{}, invalid("%s must be an object", field)
	}
	allowed := map[string]bool{}
	for _, name := range resourceFields {
		allowed[name] = true
	}
	unknown := unsupportedFields(record, allowed)
	if len(unknown) != 0 {
		return worldcut.ResourceIdentity{}, invalid(
			"%s contains unsupported field(s): %v",
			field,
			unknown,
		)
	}
	values := make([]string, 0, len(resourceFields))
	for _, name := range resourceFields {
		text, ok := record[name].(string)
		if !ok || text == "" {
			return worldcut.ResourceIdentity{}, invalid(
				"%s.%s must be a non-empty string",
				field,
				name,
			)
		}
		values = append(values, text)
	}
	return worldcut.ResourceIdentity{
		Provider: values[0],
		Account:  values[1],
		Kind:     values[2],
		Key:      values[3],
	}, nil
}

func unsupportedFields(record map[string]any, allowed map[string]bool) []string {
	var unknown []string
	for key := range record {
		if !allowed[key] {
			unknown = append(unknown, key)
		}
	}
	sort.Strings(unknown)
	return unknown
}
